### PR TITLE
services/horizon/db2/history: Improve /trades DB queries performance

### PR DESCRIPTION
**Pull request by bartekn**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ x I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [x] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

This commit improves performance of queries generated by `TradesQ` which is used in `/trades` endpoint.

First, a quick fix. A new multicolumn index was added on `base_asset_id, counter_asset_id, history_operation_id, order`. It supports `TradesForAssetPair` queries and improved the speed of such queries from ~5s on average in SDF cluster to a few ms. Adding the index took around 2 minutes in SDF cluster (with full history).

Second, I updated queries that contain `base_* = X OR counter_* = X` conditions to `UNION`s. Specifically, `ForAccount` and `ForOffer`. Previous query was using a wrong index (`htrd_pid` which required a lot of filtering) and adding new indices (`base_` + `history_operation_id` + `order`) didn't really help (I believe because of `OR` statement). New query decreased execution times down to ms.

### Known limitations

* `github.com/Masterminds/squirrel` does not support `UNION` queries so it's built manually.
* New tests are using `scc` scenarios. A larger refactor is required to support mocking.

### CODE REVIEW

File: services/horizon/internal/db2/history/trade.go [hunk 1]
```diff
@@ -120,42 +121,101 @@ func (q *TradesQ) Page(page db2.PageQuery) *TradesQ {
...
+			q.Err = errors.New("error building a firstSelect query")
+			return q
+		}
+		secondSQL, secondArgs, err := secondSelect.ToSql()
+		if err != nil {
+			q.Err = errors.New("error building a secondSelect query")
+			return q
+		}
+
+		q.rawSQL = fmt.Sprintf("(%s) UNION (%s) ", firstSQL, secondSQL)
```
Comment (tamirms) [inline services/horizon/internal/db2/history/trade.go:152]: take a look at `TransactionByHash()` to see an alternative way to construct the `UNION` query using squirrel

File: services/horizon/internal/db2/history/trade.go [hunk 2]
```diff
@@ -120,42 +121,101 @@ func (q *TradesQ) Page(page db2.PageQuery) *TradesQ {
...
+			panic("Invalid order")
+		}
+		q.rawSQL = q.rawSQL + fmt.Sprintf("LIMIT %d", page.Limit)
+		// Reset sql so it's not used accidentally
+		q.sql = sq.SelectBuilder{}
+	} else {
+		q.sql = q.appendOrdering(q.sql, op, idx, page.Order)
+	}
+
+	q.sql = q.sql.Limit(page.Limit)
```
Comment (tamirms) [inline services/horizon/internal/db2/history/trade.go:171]: should this be moved to the `else` clause?

Comment (bartekn) [reply services/horizon/internal/db2/history/trade.go]: Good catch, fixed in https://github.com/stellar/go/pull/2869/commits/013fdcbf62d3e64c0ebcda6e2dab39ba1697bdbb.

File: services/horizon/internal/db2/history/trade.go (hunk 1 above)
Comment (bartekn) [reply services/horizon/internal/db2/history/trade.go:152]: This was my first approach but it doesn't work when you want to add `ORDER BY` and/or `LIMIT` to `UNION` (requires brackets).

**Comment by ire-and-curses** [comment]
This is cool! What's the overhead for migration to add the index look like for full pubnet history?

**Comment by bartekn** [comment]
@ire-and-curses I included this in the PR summary:
> Adding an index took around 2 minutes in SDF cluster (with full history).

The new index is around ~1GB as far as I can remember. Can confirm tomorrow.

**Comment by ire-and-curses** [comment]
Super.

**Pull request merged**